lib/pbench/server/unpack_tarballs.py

```diff
@@ -100,34 +65,39 @@ def unpack_tarballs(self, min_size: float, max_size: float) -> Results:
         Returns:
             Results tuple containing the counts of Total and Successful tarballs.
         """
-        tarlist = [
-            tarball
-            for tarball in self.config.ARCHIVE.glob(f"*/{self.LINKSRC}/*.tar.xz")
-            if min_size <= Path(tarball).stat().st_size < max_size
-        ]
-
-        ntotal = nsuccess = 0
-        file_tree = FileTree(self.config, self.logger)
-
-        for tarball in sorted(tarlist):
-            ntotal += 1
-
-            try:
+        datasets = self.sync.next(Operation.UNPACK)
+        tarlist: list[Target] = []
+        for d in datasets:
+            t = Metadata.getvalue(d, Metadata.TARBALL_PATH)
+            if t:
                 try:
-                    tb = Path(tarball).resolve(strict=True)
+                    p = Path(t).resolve(strict=True)
+                    s = p.stat().st_size
                 except FileNotFoundError as exc:
-                    self.logger.error(
-                        "{}: Tarball link, '{}', does not resolve to a real location: {}",
+                    self.logger.exception(
+                        "{}: Tarball '{}' does not resolve to a file: {}",
                         self.config.TS,
-                        tarball,
+                        t,
                         exc,
                     )
-                    raise
-                self.unpack(tb, file_tree)
-                self.update_symlink(tb)
-            except Exception:
-                continue
+                    continue
+                except Exception:
+                    self.logger.exception("Unexpected exception on {}", t)
+                    continue
+
+                if min_size <= s < max_size:
+                    tarlist.append(Target(d, p))
 
+        ntotal = nsuccess = 0
+
+        for tarball in sorted(tarlist, key=lambda e: str(e.tarball)):
+            ntotal += 1
+            self.unpack(tarball)
+            self.sync.update(
+                dataset=tarball.dataset,
+                did=Operation.UNPACK,
+                enabled=[Operation.COPY_SOS, Operation.INDEX],
+            )
```

Comment:
It feels kind of weird to have the `enabled` values hardcoded in each of the `sync.update()` calls sprinkled throughout the code.  It seems like there ought to be a table somewhere, like the dataset state transition table, which defines each of the `enabled` lists based on the `did` operation.  That would centralize the flow in a place where we could examine the whole thing, and where we could make any changes required without combing through the code.  And, it might allow the `update()` interface to be simplified (although we would have to ponder that -- the value of the `enabled` parameter could be predicted by `update()` based on the `did` parameter and the table, but we might not want to make the table available to `update()`, in which case we would want to still pass the value in).

Reply:
I did it that way to allow for flexibility when I started designing it on the fly.

You're right that the actual transitions are well defined, and I probably *could* remove `enabled` and infer the new enablements from the `did`. I never went back to think about that, and I need to consider whether I really want to take that on now during review as it'd make a fair amount of (admittedly localized) churn. I'll give this some thought overnight, I guess.

Reply:
The point about the churn is a valid one.  Perhaps it would be better to make that change in a subsequent PR.  (Although, once I finish with my first review pass, I should be able to handle this change, provided it's in a separate commit....)